lib/rules/no-restricted-imports.js

```diff
@@ -121,31 +139,128 @@ module.exports = {
+                isEverythingImported(node);
+        }
+
+        /**
+         * Check if the given objectName is restricted given a list of restrictedObjectNames.
+         * @param {[string]} objectNames - array of object names that are being imported
+         * @param {[string]} restrictedObjectNames - array of object names that are restricted for this import
+         * @returns {boolean} whether the objectName is restricted
+         * @private
+         */
+        function isRestrictedObject(objectNames, restrictedObjectNames) {
+            const restrictedObjectMatches = restrictedObjectNames.reduce((accumulator, restrictedObjectName) => {
+                if (objectNames.indexOf(restrictedObjectName) > -1) {
+                    accumulator.push(restrictedObjectName);
+                }
+                return accumulator;
+            }, []);
+
+            return restrictedObjectMatches.length > 0;
+        }
+
+        /**
          * Check if the given name is a restricted path name.
-         * @param {string} name name of a variable
+         * @param {string} name - path of the import
+         * @param {[string]} [objectNames] - array of object names that are being imported
          * @returns {boolean} whether the variable is a restricted path or not
          * @private
          */
-        function isRestrictedPath(name) {
-            return Object.prototype.hasOwnProperty.call(restrictedPathMessages, name);
+        function isRestrictedPath(name, objectNames) {
+            let isRestricted = false;
+
+            if (Object.prototype.hasOwnProperty.call(restrictedPathMessages, name)) {
+                if (restrictedPathMessages[name].objectNames) {
+                    isRestricted = isRestrictedObject(objectNames, restrictedPathMessages[name].objectNames);
+                } else {
+                    isRestricted = true;
+                }
+            }
+
+            return isRestricted;
+        }
+
+        /**
+         * Check if the given name is a restricted path name.
+         * @param {string} name - path of the import
+         * @returns {boolean} whether the variable is a restricted pattern or not
+         * @private
+         */
+        function isRestrictedPattern(name) {
+            return restrictedPatterns.length > 0 && restrictedPatternsMatcher.ignores(name);
         }
 
         return {
             ImportDeclaration(node) {
+
                 if (node && node.source && node.source.value) {
                     const importName = node.source.value.trim();
+                    const importObjectNames = node.specifiers.map(specifier => {
+                        if (specifier && specifier.type === "ImportDefaultSpecifier") {
+                            return "default";
+                        }
+                        return (specifier && specifier.imported && specifier.imported.name) ||
+                               (specifier && specifier.local && specifier.local.name) || null;
```

Comment:
Are all of these null-checks necessary? To me it seems like `specifier.imported.name` will always exist, so this line could just be `return specifier.imported.name;`.

Reply:
From my original testing, `specifier.imported` is only populated when using an alias.  I will double check this again.

Worst case, I think I should be able to remove the null check on `specifier`.

Reply:
You are right again!  I don't need any additional checks anymore, I can just return `specifier.imported.name`.  I think adding logic for the `default` and `*` case made these these checks unnecessary.